src/EFCore.PG/Storage/Internal/Mapping/NpgsqlCharacterTypeMapping.cs

```diff
@@ -64,5 +65,11 @@ protected override void ConfigureParameter(DbParameter parameter)
 
             base.ConfigureParameter(parameter);
         }
+
+        static bool EqualsWithoutTrailingWhitespace(string a, string b)
+            => a.AsSpan().TrimEnd().SequenceEqual(b.AsSpan().TrimEnd());
```

Comment:
I checked and the call to `AsSpan()` actually returns an empty `ReadOnlySpan` on null, so we don't need the checks for exception safety. 

I also looked around and couldn't find any documentation saying that these value comparers are meant to replicate SQL nullability (e.g. `null != null`). 

So, my guess is that (like you suggested on #820) the nulls are handled elsewhere, which would make sense so that `UseRelationalNulls` gets handled centrally.

Reply:
I'd try to look at what mappings in EF Core actually do (this kind of thing isn't likely to be documented)... Whatever you think is right.